[JIT] Build mechanism for custom operators

goldsborough opened the pull request:
This is the last step in the custom operator implementation: providing a way to build from C++ and Python. For this I:

1. Created a `FindTorch.cmake` taken largely from @ebetica with a CMake function to easily create simple custom op libraries
2. Created a ` torch/op.h` header for easy inclusion of necessary headers,
3. Created a test directory `pytorch/test/custom_operator` which includes the basic setup for a custom op. 
    1. It defines an op in `op.{h,cpp}`
    2. Registers it with the JIT using `RegisterOperators`
    3. Builds it into a shared library via a `CMakeLists.txt`
    4. Binds it into Python using a `setup.py`. This step makes use of our C++ extension setup that we already have. No work, yey!

The pure C++ and the Python builds are separate and not coupled in any way.

@zdevito @soumith @dzhulgakov

goldsborough commented:
Tested in Python, now ready for review! Also added better docstrings to automatically bound ops:

<img width="817" alt="screen shot 2018-08-06 at 15 30 37" src="https://user-images.githubusercontent.com/6429851/43744245-020c0ec2-998e-11e8-876e-547c3065d9b4.png">

goldsborough committed:
Create torch/op.h for custom operators

Create CMake setup for custom ops

Fixes to enable custom op build

Figure out Python integration

Fix Windows

Fix python test for custom op and give bound operators better docstrings

goldsborough commented on test/custom_operator/test.py:
```diff
@@ -0,0 +1,9 @@
+# All this does is load the dynamic library which runs the global op registration code.
+import custom_op
+
+output = torch.ops.custom.op(torch.ones(5), 2.0, 3)
```
I should add a test here that uses the op in a script module

dzhulgakov commented on test/custom_operator/op.cpp:
```diff
@@ -0,0 +1,37 @@
+
+torch::RegisterOperators registry("custom::op", &custom_op);
+
+#if defined(WITH_PYTHON_OP_BINDINGS)
```
you don't need to create python module. If you create the shared library that works through the registration mechanism, it's sufficient to just load it into memory and it can be done ctypes.CDLL

actually you can verbatim copy the mechanism Caffe2 uses as it's exactly the same problem. See:
* https://github.com/pytorch/pytorch/blob/HEAD/caffe2/python/dyndep.py
* https://github.com/pytorch/pytorch/blob/HEAD/caffe2/python/extension_loader.py (handles RTLD_GLOBAL)
* pytorch already handles RTLD_GLOBAL for the main _C.so: https://github.com/pytorch/pytorch/blob/HEAD/torch/__init__.py#L75

The only benefit of setup.py I can see is that setuptools manages the install path for the package. However, it depends on what's the target workflow you imagine for people developing c++ extensions. Would they install it to system paths? Or would they just drop the .so in the current model folder and load it directly?

goldsborough replied:
Thanks for the cdll links, looks awesome! For C++ extensions setuptools helped us getting the paths to Python.h.

I think because these JIT/script extensions are intended more for production use cases, compared to non-JIT C++ extensions, the pure .so path without setuptools might be the one to advertise

dzhulgakov commented on test/custom_operator/op.cpp:
```diff
@@ -0,0 +1,37 @@
+  return output;
+}
+
+torch::RegisterOperators registry("custom::op", &custom_op);
```
put it into anonymous namespace? also it might be good to have a macro like REGISTER_CPU_OPERATOR in Caffe2 that takes care of the temporary variable name for you (again, we should just move that part to ATen/core and use verbatim probably)

goldsborough replied:
I think @zdevito and I would prefer fewer macros. The code looks pretty simple and readable to me :)

dzhulgakov commented on torch/csrc/jit/custom_operator.h:
```diff
@@ -204,21 +199,54 @@ Operator createOperator(
+/// so in the global scope when a `RegisterOperators` object is assigned to a
+/// static variable. Also handles registration of user-defined, "custom"
+/// operators.
+struct TORCH_API RegisterOperators {
```
it's a bit weird that it's plural but used usually to register a single op

goldsborough replied:
The original code using this, e.g. `register_aten_ops.cpp`, looks more like this:
```
RegisterOperators reg({
  Operator(
      "aten::RoiPooling2d_backward(Tensor input, Tensor rois, int pooledHeight, int pooledWidth, float spatialScale, Tensor gradOutput, Tensor argmaxes) -> Tensor",
      [](Stack & stack) {
          autograd::profiler::RecordFunction record("RoiPooling2d_backward");
          auto result = at::RoiPooling2d_backward(
              (std::move(peek(stack, 0, 7))).toTensor(),
              (std::move(peek(stack, 1, 7))).toTensor(),
              (std::move(peek(stack, 2, 7))).toInt(),
              (std::move(peek(stack, 3, 7))).toInt(),
              (std::move(peek(stack, 4, 7))).toDouble(),
              (std::move(peek(stack, 5, 7))).toTensor(),
              (std::move(peek(stack, 6, 7))).toTensor()
          );
          drop(stack, 7);
          pack(stack, std::move(result));
          return 0;
      }
  ),
  Operator(
      "aten::RoiPooling2d_forward(Tensor input, Tensor rois, int pooledHeight, int pooledWidth, float spatialScale) -> (Tensor, Tensor)",
      [](Stack & stack) {
          autograd::profiler::RecordFunction record("RoiPooling2d_forward");
          auto result = at::RoiPooling2d_forward(
              (std::move(peek(stack, 0, 5))).toTensor(),
              (std::move(peek(stack, 1, 5))).toTensor(),
              (std::move(peek(stack, 2, 5))).toInt(),
              (std::move(peek(stack, 3, 5))).toInt(),
              (std::move(peek(stack, 4, 5))).toDouble()
          );
          drop(stack, 5);
          pack(stack, std::move(result));
          return 0;
      }
  ),
  Operator(
      "aten::__and__(Tensor self, Tensor other) -> Tensor",
      [](Stack & stack) {
          autograd::profiler::RecordFunction record("__and__");
          auto result = at::__and__(
              (std::move(peek(stack, 0, 2))).toTensor(),
              (std::move(peek(stack, 1, 2))).toTensor()
          );
          drop(stack, 2);
          pack(stack, std::move(result));
          return 0;
      }
  ),
...
```

dzhulgakov commented on torch/csrc/jit/custom_operator.h:
```diff
@@ -204,21 +199,54 @@ Operator createOperator(
+/// static variable. Also handles registration of user-defined, "custom"
+/// operators.
+struct TORCH_API RegisterOperators {
+  RegisterOperators() = default;
```
why not =delete?

goldsborough replied:
I was thinking it might be useful to first declare the `RegisterOperators` object, and then call `.op(...)` on it after, e.g. if you're looping over a vector

goldsborough committed:
Dynamically load library with ctypes

goldsborough committed:
Add test for custom ops in script

goldsborough commented:
Thanks for the review! I added code to dynamically load a shared library in the way that caffe2 does it. This avoid having to make the user `.so` Python-specific in any way, and doesn't even require a `setup.py` at all. I also added a test for using a registered op from a `@torch.jit.script`-ed function.

goldsborough committed:
Uncomment tests

goldsborough commented:
One last TODO is to try the end-to-end flow of defining a script module in C++ that uses a custom operator registered inside Python, then exporting that to something that is loaded from C++, and then executing the module from C++

dzhulgakov commented on cmake/Modules/FindTorch.cmake:
```diff
@@ -0,0 +1,75 @@
+# FindTorch
```
I'm not a CMake expert. But I think the recommended way is to produce -config.cmake file instead of FindXXX.cmake.

This is how Caffe2 does it: https://github.com/pytorch/pytorch/blob/master/cmake/Caffe2Config.cmake.in https://github.com/pytorch/pytorch/commit/a1518b780138400af98df6ea0096f39f2481191d#diff-dcfb71fad8d9192fd423940106d63b66

Resident CMake experts: @Yangqing @orionr @ezyang

dzhulgakov replied:
Checked with @Yangqing offline. -config.cmake is indeed a recommended way - you can follow the C2 example.

Or see https://gitlab.kitware.com/cmake/community/wikis/doc/tutorials/How-to-create-a-ProjectConfig.cmake-file

goldsborough replied:
Ok I'll try this

dzhulgakov commented on test/custom_operator/test.py:
```diff
@@ -0,0 +1,12 @@
+
+library_path = os.path.abspath('build/libcustom_op.so')
+torch.ops.load_library(library_path)
+assert torch.ops.loaded_libraries == {library_path}
```
why not `in` (there might be other libs loaded)

dzhulgakov commented on test/test_jit.py:
```diff
@@ -6118,19 +6118,30 @@ def test_passing_unknown_kwargs(self):
+    def test_script_graph_contains_custom_op(self):
+        @torch.jit.script
+        def func(x):
+            return torch.ops.aten.relu(x)
```
is it actually a proper custom op?

goldsborough replied:
It's not declared in a file separate from the repo, but the way these aten ops are registered is exactly the way we register user ops (it's that long list of `Operator` definitions I pasted in an earlier comment above), so the relevant code paths should be tested

apaszke replied:
This is very confusing for me. Why would anyone want to use `torch.ops` instead of some handles to their own shared library they import? I can't see how forcing everyone into a single namespace is a good thing.

goldsborough replied:
It’s so that users don’t have to write Python bindings. We generate them automatically. See torch/_ops.py

dzhulgakov reviewed (commented):
Looks good to me, but please double check the CMake part

zdevito reviewed (approved):
The JIT changes look good. I don't know enough about CMake best practices to comment there.

goldsborough commented on test/custom_operator/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
+project(custom_op)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../../cmake/Modules")
```
Users have to add the path to the `FindTorch.cmake` file, I don't think there's a way around this

goldsborough committed:
Create TorchConfig.cmake

goldsborough committed:
Outdated comment and nit in custom_operator/test.py

goldsborough commented:
I think if tests pass I will land this diff. I know the CMake stuff can be improved, but this diff has code changes unrelated to the cmake. I will then create a new PR that integrates the custom operator tests with CI, in which we can discuss the proper way of doing the TorchConfig.cmake. I spent a little time looking at how Caffe2Config.cmake works, but there may be some open questions
